typify/shared/src/main/scala/typify/GCursor.scala

```diff
@@ -0,0 +1,339 @@
+package typify
+
+import cats.{Functor, Id}
+import cats.data.{NonEmptyList, NonEmptyVector, Validated, ValidatedNel}
+import cats.syntax.either._
+import cats.syntax.functor._
+import cats.syntax.list._
+import cats.syntax.semigroup._
+import cats.syntax.validated._
+import scala.annotation.tailrec
+import scala.language.dynamics
+import shapeless.tag
+
+/** A cursor representing a view of a specific value of type `A` in an overall value of type `T`
+  *
+  * @constructor Constructor
+  * @tparam T The type of the overall root value
+  * @tparam A The type of the current value
+  * @param value The current value
+  * @param lastOp The last operation performed to produce this cursor
+  */
+sealed abstract class GCursor[T, A](val value: A, val lastOp: CursorOp) extends Dynamic { self0 =>
+  import GCursor._
+
+  /** The type of the current focus, may differ from `A` if additional data is needed to support moving the focus */
+  type Focus
+
+  /** The type of the last cursor that produced this cursor */
+  type Last <: GCursor[T, _]
+
+  /** The root cursor focused on a value of type `T` */
+  def root: GCursor.Top[T]
+
+  /** The current focus */
+  def focus: Focus
+
+  /** The last cursor that produced this cursor */
+  def last: Option[Last]
+
+  private final type Self = GCursor.AuxL[T, A, Focus, Last]
+  @inline private final val self: Self = self0
+
+  /** The type of a new cursor produced by this cursor
+    *
+    * @tparam NewValue The type of the new value
+    * @param NewFocus The type of the new focus
+    */
+  final type Next[NewValue, NewFocus] = GCursor.AuxL[T, NewValue, NewFocus, Self]
+
+  /** The history of `CursorOp`s that led to this cursor */
+  @inline final lazy val history: Vector[CursorOp] = {
+    @tailrec def go(res: Vector[CursorOp], next: Option[GCursor[T, _]]): Vector[CursorOp] = next match {
+      case Some(c) => go(res :+ c.lastOp, c.last)
+      case None => res
+    }
+
+    go(Vector(), Some(this))
+  }
+
+  /** Create a failed `CursorHistory` */
+  private final def failedHistory(failedOp: CursorOp): CursorHistory =
+    CursorHistory(history, Some(NonEmptyVector.of(failedOp)))
+
+  @inline final private def id[B](b: B): Id[B] = b
+  @inline final private def dup[B](b: B): Id[(B, B)] = (b, b)
```

Comment:
this is 1/2 monoid and 1/2 comonoid - worth thinking about b/c those are precursors to monad/comonad if you don't have either already